### ISSUE TITLE
SSH connection timeout 

### ISSUE THREAD

**Issue by loxal** [hetznercloud/terraform-provider-hcloud#29]
I run into a timeout with hcloud_server:

> hcloud_server.node: Still creating... (5m0s elapsed)
> hcloud_server.node (remote-exec): Connecting to remote host via SSH...
> hcloud_server.node (remote-exec):   Host: 39.69.145.168
> hcloud_server.node (remote-exec):   User: root
> hcloud_server.node (remote-exec):   Password: false
> hcloud_server.node (remote-exec):   Private key: false
> hcloud_server.node (remote-exec):   SSH Agent: false
> hcloud_server.node (remote-exec):   Checking Host Key: false
> hcloud_server.node: Still creating... (5m10s elapsed)
> 
> Error: Error applying plan:
> 
> 1 error(s) occurred:
> 
> * hcloud_server.node: timeout - last error: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
> 
> Terraform does not automatically rollback in the face of errors.
> Instead, your Terraform state file has been partially updated with
> any resources that successfully completed. Please address the error
> above and apply again to incrementally change your infrastructure.

**Comment by LKaemmerling** [comment]
Hello,

this should be a local issue, please check your firewall. 

I have tested it locally and it works correctly on all versions.

**Comment by loxal** [comment]
Locally it works too. Running the Terraform locally, I provisioned a Hetzner machine on which I wanted to execute Terraform as root user which returned the error you see. I have the same problem running this with a hashicorp/terraform:light container. The Hetzner server that I provisioned is a Debian machine that does not have any visible firewall rules.

**Comment by LKaemmerling** [comment]
Okay, then it is a really easy issue :) The Server you create does not have the ssh key, which you want to use. 

`unable to authenticate, attempted methods [none publickey], no supported methods remain` Means, there is an SSH Connection, but the server could not be authenticated.

**Comment by loxal** [comment]
I think the server does not need any SSH key because the authentication happens not via SSH(?). But still I created some random key, and also copied one of the private keys I use. You can easily reproduce the problem on a freshly provisioned Debian or Ubuntu machine

**Comment by LKaemmerling** [comment]
Can you send me your terraform description per E-Mail to lukas.kaemmerling@hetzner-cloud.de? 

Terraform checks if a server was created via SSH. So there is an SSH Connection. 

I have tested it with a fresh ubuntu machine and does not get this behavior. The IP you showed in the log above is not an IP from Hetzner.

**Comment by loxal** [comment]
> This ist the TF definition I want to materialize:
> 
> variable "tenant" {
>   type = "string"
>   default = "node"
>   description = "Customer Name"
> }
> 
> variable "password" {
>   type = "string"
>   default = "invalid dummy"
>   description = "Default Password"
> }
> 
> variable "hetzner_cloud_analyze_law" {
>   type = "string"
>   default = "invalid dummy"
>   description = "Analyze Law API Token"
> }
> 
> locals {
>   hcloud_token = "${var.hetzner_cloud_analyze_law}"
>   //    server_image = "ubuntu-18.04"
>   server_image = "debian-9"
>   datacenter_prefix = "nbg1"
>   tenant = "${var.tenant == "node" ? terraform.workspace : var.tenant}"
>   //  tenant = "${var.tenant}"
> }
> 
> provider "hcloud" "Hetzner" {
>   token = "${var.hetzner_cloud_analyze_law}"
> }
> 
> resource "hcloud_server" "node" {
>   name = "${terraform.workspace}-${local.tenant}-${count.index}"
>   count = "1"
>   datacenter = "${local.datacenter_prefix}-dc3"
>   //  datacenter = "fsn1-dc14"
>   image = "${local.server_image}"
>   server_type = "cx31-ceph"
>   ssh_keys = [
>     "alex",
>     "amer",
>     "bachka",
>     "minion",
>   ]
> 
>   provisioner "remote-exec" "install" {
>     inline = [
>       "apt-get install -y curl software-properties-common",
>     ]
>   }
> }

I run it on the machine that I materialized using the same definition on my localhost. I copied the 'minion" user's SSH private & public keys to this Hetzner machine. From this Hetzner machine I can **ssh root@159.69.145.168**. After the the above described failure, I can even ssh from the executing Hetzner machine to the newly half-created Hetzner machine. But still the provisioning does not succeed. The IP from my previous post was purposely altered.  

So basically the Hetzner machine where I execute Terraform does have the "minion" SSH key.

**Comment by loxal** [comment]
@LKaemmerling 
![screenshot from 2018-09-13 12-17-27](https://user-images.githubusercontent.com/87507/45482689-0476e400-b74f-11e8-878d-3a87203bc4ff.png)

**Comment by LKaemmerling** [comment]
The error is your configuration.

The log file shows the following:
`hcloud_server.node (remote-exec): Private key: false`

So you do not specify the private key, which should be used for the connection.
See: https://www.terraform.io/docs/provisioners/connection.html

Your configuration should contain the following:

```
provisioner "remote-exec" "install" {
    connection {
        type     = "ssh"
        user     = "root"
        private_key = "{PATH_TO_YOUR_PRIVATE_KEY}"
    }

    inline = [
      "apt-get install -y curl software-properties-common",
    ]
  }
```

The next problem, would be, that you get a error message like 
```

hcloud_server.rdns1 (remote-exec): E: Could not get lock /var/lib/dpkg/lock - open (11: Resource temporarily unavailable)
hcloud_server.rdns1 (remote-exec): E: Unable to lock the administration directory (/var/lib/dpkg/), is another process using it?

```
This is because, the server is in the booting phase. You should put a `sleep 10` before the `apt-get install`. Now it should work without any problems :)

**Comment by loxal** [comment]
Yeah, I solved the _apt-get lock problem_ with exactly this approach :) However only Ubuntu requires this hacky solution. Debian boots and installs everything fine without sleeping. Maybe because it boots faster...

However, in GitLab CI, using _hashicorp/terraform:light_ container I am doing this

> $ cp -r ssh ~/.ssh && chmod -R 600 ~/.ssh && ls -lash ~/.ssh
> total 16
>      4 drw-------    2 root     root        4.0K Sep 13 11:36 .
>      4 drwx------    1 root     root        4.0K Sep 13 11:36 ..
>      4 -rw-------    1 root     root        1.6K Sep 13 11:36 id_rsa
>      4 -rw-------    1 root     root         391 Sep 13 11:36 id_rsa.pub
> $ ssh -o StrictHostKeyChecking=no root@159.69.195.109
> [LOGGING TO A HETZNER MACHINE WITH THIS KEY WORKS FINE HERE]

but later I run into 

> Destroy complete! Resources: 0 destroyed.
> $ terraform apply -auto-approve
> data.google_dns_managed_zone.analyze-law: Refreshing state...
> hcloud_server.node: Creating...
>   datacenter:   "" => "nbg1-dc3"
>   image:        "" => "debian-9"
>   ipv4_address: "" => "<computed>"
>   ipv6_address: "" => "<computed>"
>   ipv6_network: "" => "<computed>"
>   keep_disk:    "" => "false"
>   location:     "" => "<computed>"
>   name:         "" => "default-default-0"
>   server_type:  "" => "cx31-ceph"
>   ssh_keys.#:   "" => "4"
>   ssh_keys.0:   "" => "alex"
>   ssh_keys.1:   "" => "amer"
>   ssh_keys.2:   "" => "bachka"
>   ssh_keys.3:   "" => "minion"
>   status:       "" => "<computed>"
> hcloud_server.node: Still creating... (10s elapsed)
> hcloud_server.node: Provisioning with 'file'...
> hcloud_server.node: Still creating... (20s elapsed)
> hcloud_server.node: Still creating... (30s elapsed)
> hcloud_server.node: Still creating... (40s elapsed)

Maybe I am going to try with a regular user. This is quite strange as exactly the same stuff works well on my localhost. 

Btw, we just paid a 52 page long Hetzner bill... ;)

**Comment by LKaemmerling** [comment]
@loxal hey this is cool :) my longest bill was more than 300 pages. If you run into more or another issues, just write me a mail, hopefully, I can help you.

**Comment by loxal** [comment]
I assume this is the problem

> root@dev-dev-0 ~ # ssh-keygen -t rsa
> Generating public/private rsa key pair.
> Enter file in which to save the key (/root/.ssh/id_rsa): 
> /root/.ssh/id_rsa already exists.
> Overwrite (y/n)? y
> Enter passphrase (empty for no passphrase): 
> Enter same passphrase again: 
> Your identification has been saved in /root/.ssh/id_rsa.
> Your public key has been saved in /root/.ssh/id_rsa.pub.
> The key fingerprint is:
> SHA256:PFi22QBj9tUzxviZIbzqEusaWqxCN5MpFGEqNXqecr0 root@dev-dev-0
> The key's randomart image is:
> +---[RSA 2048]----+
> | o+   =  ..+     |
> |o+ . o + .+ B    |
> |+ o     =  = *   |
> |.+ o   = =. +    |
> |o + + . S..      |
> | = B .. ..       |
> |. o E  +         |
> |.  + .o .        |
> | .o .o..         |
> +----[SHA256]-----+
> **root@dev-dev-0 ~ # ssh-add -l**
> **Could not open a connection to your authentication agent.**

**Comment by loxal** [comment]
This ist thing is quite frustrating and time consuming. Maybe you could create an example declaration that creates a server and pushes itself to this server in order to be executed there. If this sample works, it would enable CI builds to work. I.e. this GitLab declaration times out for the same reason. You do need GitLab to test it.

```
provision:
  image:
    name: hashicorp/terraform:light
    entrypoint:
    - /usr/bin/env
  stage: provision
  script:
  - cd ops/al-provisioning
  - terraform init
  - terraform plan
  - terraform destroy -auto-approve
  - terraform apply -auto-approve
```
Just test it here:
`docker run -w /app -v $(pwd):/app --entrypoint /usr/bin/env -it hashicorp/terraform:light sh`

**Comment by loxal** [comment]
Just execute...

> docker run -w /app -v $(pwd):/app --entrypoint /usr/bin/env -it hashicorp/terraform:light sh

...from your Terraform directory and you can reproduce this error. This is how many people would run Terraform on their CIs. Therefore this error might affect many people. Please let me know if this worked for as I cannot imagine that this would work.

**Comment by LKaemmerling** [comment]
What is your configuration now?

**Comment by loxal** [comment]
I sent you my exact configuration via e-mail.

**Comment by loxal** [comment]
The declaration that I have sent you, contains non-hcloud parts, but those parts are not the trouble makers. I checked and they work absolutely fine. As I also said, everything runs absolutely fine on my Linux computer. This issue has nothing to do with missing SSH keys as I 600-copied them everywhere I run Terraform.

You can definitely reproduce this issue running this in any of your Terraform directories that contain **hcloud_server**:
```
docker run -w /app -v $(pwd):/app --entrypoint /usr/bin/env -it hashicorp/terraform:light sh
rm -rf .terraform
terraform init
terraform apply
```

This is crucial for integrating Terraform in any modern CI environment.

**Comment by LKaemmerling** [comment]
This must be an issue with your configuration.

I have setup a basic cx11 instance, installed docker, generated an ssh key, create a directory where i pasted the following configuration to it:
```hcl
provider "hcloud" {
  token = "${var.hcloud_token}"
}

resource "hcloud_server" "rdns1" {
  name        = "rdns-1-123"
  server_type = "cx11"
  image       = "debian-9"
  datacenter  = "fsn1-dc8"
  ssh_keys    = ["93534", "your_email@example.com"]

  provisioner "remote-exec" "install" {
    connection {
        type     = "ssh"
        user     = "root"
        private_key = "${file("id_rsa")}"
    }

    inline = [
      "sleep 10",
      "apt-get install -y curl software-properties-common",
    ]
  }
}
```

Then i run the command from above:
```
docker run -w /app -v $(pwd):/app --entrypoint /usr/bin/env -it hashicorp/terraform:light sh
rm -rf .terraform
terraform init
terraform apply
```

And it created the server and installed the `software-properties-common` without any problems.

**Comment by loxal** [comment]
Thanks for keeping patience with this issue. Within _hcloud_server_ I had a some file provisioners that prevented further execution. I had to add (in my case) 
```
    connection {
# other values are implicit defaults 
        private_key = "${file("ssh/id_rsa")}"
    }
```
not only to the _hcloud_server_ resource but also the file provisioners within this resource. This **solved** my issue. Thanks again :+1: 
Apparently it worked locally because on desktop machines usually an SSH agent runs behind the scenes.
